### PR TITLE
Feature/staking service

### PR DESCRIPTION
**Pull request by liayoo**
- Revamp _deposit, _withdraw. Now we have staking services with _stake and _unstake native functions
  - /deposit/${service_name}/${user_addr}/${key} -> /staking/${service_name}/${user_addr}/${staking_key}/stake/${record_id}
  - /withdraw/${service_name}/${user_addr}/${key} -> /staking/${service_name}/${user_addr}/${staking_key}/unstake/${record_id}
- Replace deposit_accounts with service_accounts (#239)
  - /deposit_accounts/${service_name}/${user_addr}/value -> /service_accounts/staking/${service_name}/${user_addr}|${staking_key}/balance
- Total staking balance of an app is tracked at /staking/${service_name}/balance_total
- User's staking lockup expiration time is tracked at /staking/${service_name}/${user_addr}/${staking_key}/expire_at
- Add minimal implementation for manage_app and _createApp (service consistency & reusability features are coming soon!)
- Add tool scripts for staking

### CODE REVIEW

File: blockchain/block.js [hunk 1]
```diff
@@ -226,6 +227,29 @@ class Block {
...
+    const thirdTxBody = {
+      nonce: -1,
+      timestamp,
+      operation: {
+        type: 'SET_VALUE',
+        ref: `${PredefinedDbPaths.MANAGE_APP}/${PredefinedDbPaths.CONSENSUS}/` +
+            `${PredefinedDbPaths.MANAGE_APP_CREATE}/${timestamp}`,
+        value: {
+          [PredefinedDbPaths.MANAGE_APP_CONFIG_ADMIN]: {
+            AI_NETWORK: true
```
Comment (platfowner) [inline blockchain/block.js:240]: It's hard to find this variable's definition?

File: db/functions.js [hunk 2]
```diff
@@ -453,34 +454,81 @@ class Functions {
...
-  _deposit(value, context) {
-    const service = context.params.service;
+  _createApp(value, context) {
+    const appName = context.params.app_name;
+    const recordId = context.params.record_id;
+    const timestamp = context.timestamp;
+    const transaction = context.transaction;
+    const auth = context.auth;
+    const resultPath = this.getCreateAppResultPath(appName, recordId);
+    const lockupDurationKey = `${PredefinedDbPaths.MANAGE_APP_CONFIG_SERVICE}.` +
```
Comment (platfowner) [inline db/functions.js:464]: It looks like _createApp() knows about the structure of the service config.
Could it be made independent of any app or service details?

Comment (platfowner) [reply db/functions.js:464]: It can be done by blind-copying the whole service config to the path?

File: db/functions.js [hunk 3]
```diff
@@ -489,38 +537,35 @@ class Functions {
...
+    const resultPath = this.getStakingUnstakeResultPath(serviceName, user, stakingKey, recordId);
+    const expireAt = this.db.getValue(this.getStakingExpirationPath(serviceName, user, stakingKey));
     if (expireAt > executedAt) {
       // Still in lock-up period.
       this.saveAndSetExecutionResult(context, resultPath, FunctionResultCode.IN_LOCKUP_PERIOD);
       return;
     }
-    if (service === PredefinedDbPaths.CONSENSUS) {
-      // Reject withdrawing consensus deposits if it reduces the number of validators to less than
+    if (serviceName === PredefinedDbPaths.CONSENSUS) {
```
Comment (platfowner) [inline db/functions.js:556]: Can we make _unstake() ignorant of the consensus service logic?
I think easiest way is just dropping this block?

File: genesis-configs/base/genesis_rules.json [hunk 4]
```diff
@@ -3,7 +3,7 @@
   "accounts": {
     "$user_addr": {
       "balance": {
-        ".write": "auth.fid === '_transfer' || auth.fid === '_deposit' || auth.fid === '_withdraw' || auth.fid === '_pay' || auth.fid === '_claim'"
+        ".write": "auth.fid === '_transfer' || auth.fid === '_stake' || auth.fid === '_unstake' || auth.fid === '_pay' || auth.fid === '_claim'"
```
Comment (platfowner) [inline genesis-configs/base/genesis_rules.json:6]: Could we use simplified condition here as we're enforcing other native functions to trigger _transfer always?
`"auth.fid === '_transfer'"`
If you think it's too early, I'm fine to leave it as a todo.

File: json_rpc/index.js [hunk 5]
```diff
@@ -357,11 +357,14 @@ module.exports = function getMethods(node, p2pServer, minProtocolVersion, maxPro
 
     ain_isValidator: function(args, done) {
       const whitelisted = p2pServer.node.db.getValue(
-          `${PredefinedDbPaths.DEPOSIT_ACCOUNTS_CONSENSUS}/${PredefinedDbPaths.WHITELIST}/${args.address}`);
-      const deposit = p2pServer.node.db.getValue(
-          `${PredefinedDbPaths.DEPOSIT_ACCOUNTS_CONSENSUS}/${args.address}`);
-      const stakeValid = deposit && deposit.value > 0 &&
-          deposit.expire_at > Date.now() + ConsensusConsts.DAY_MS;
+          `${PredefinedDbPaths.CONSENSUS}/${PredefinedDbPaths.WHITELIST}/${args.address}`);
```
Comment (platfowner) [inline json_rpc/index.js:360]: Maybe we can consider extracting the `get...Path()` functions from functions.js to some common place like `common/path-util.js` to reuse them in other places as we're doing in gpt2-related projects?

Comment (platfowner) [reply json_rpc/index.js:360]: Just fine to leave it as a todo if you agree with the idea.

File: stop_servers.sh [hunk 6]
```diff
@@ -1,6 +1,5 @@
 
 BASEDIR=$(dirname "$0")
 rm -rf $BASEDIR/blockchains
```
Comment (platfowner) [inline stop_servers.sh:3]: It seems some changes were reverted by mistake: https://github.com/ainblockchain/ain-blockchain/pull/286/files#diff-5be8c0c085e331977f9996a9ee035d1c9d2ec88e70aa497381f512700c1bcb14

Could you fix it for me?

**Comment by platfowner** [comment]
Happy to see this implementation!
Let me leave some comments (esp. _createApp's independency of services / apps) and suggestions.

File: blockchain/block.js (hunk 1 above)
Comment (liayoo) [reply blockchain/block.js]: Now defined in constants.js!

File: db/functions.js (hunk 2 above)
Comment (liayoo) [reply db/functions.js:463]: Done, thanks!

File: db/functions.js (hunk 3 above)
Comment (liayoo) [reply db/functions.js]: Done!

File: genesis-configs/base/genesis_rules.json (hunk 4 above)
Comment (liayoo) [reply genesis-configs/base/genesis_rules.json]: Done! Removed other native functions from accounts and service_accounts.

File: json_rpc/index.js (hunk 5 above)
Comment (liayoo) [reply json_rpc/index.js]: Right, this was long overdue..! Moved the path functions to common/path-util.js.

File: stop_servers.sh (hunk 6 above)
Comment (liayoo) [reply stop_servers.sh]: Done!

**Comment by liayoo** [comment]
Addressed all comments, PTAL @platfowner

File: blockchain/block.js [hunk 7]
```diff
@@ -226,6 +232,28 @@ class Block {
...
+  static buildConsensusAppTx(timestamp, privateKey) {
+    const thirdTxBody = {
+      nonce: -1,
+      timestamp,
+      operation: {
+        type: 'SET_VALUE',
+        ref: getCreateAppRecordPath(PredefinedDbPaths.CONSENSUS, timestamp),
+        value: {
+          [PredefinedDbPaths.MANAGE_APP_CONFIG_ADMIN]: {
+            [AI_NETWORK]: true
```
Comment (platfowner) [inline blockchain/block.js:244]: minor: If it's just a placeholder, could you leave some comments?

File: db/functions.js [hunk 8]
```diff
@@ -453,34 +453,76 @@ class Functions {
...
+    const appName = context.params.app_name;
+    const recordId = context.params.record_id;
+    const timestamp = context.timestamp;
+    const transaction = context.transaction;
+    const auth = context.auth;
+    const resultPath = PathUtil.getCreateAppResultPath(appName, recordId);
+    const lockupDurationKey = `${PredefinedDbPaths.MANAGE_APP_CONFIG_SERVICE}.` +
+        `${PredefinedDbPaths.STAKING}.${PredefinedDbPaths.STAKING_LOCKUP_DURATION}`;
+    const lockupDurationVal = _.get(value, lockupDurationKey);
+    if (!ChainUtil.isDict(_.get(value, PredefinedDbPaths.MANAGE_APP_CONFIG_ADMIN)) ||
```
Comment (platfowner) [inline db/functions.js:466]: Now _createApp doesn't recognize 'lockupDurationVal' so would it be a good compromise:
`!ChainUtil.isDict(serviceConfig)`?

And what about moving this if-block just after 474?

File: genesis-configs/base/genesis_rules.json (hunk 4 above)
Comment (platfowner) [reply genesis-configs/base/genesis_rules.json]: Thanks!

File: json_rpc/index.js (hunk 5 above)
Comment (platfowner) [reply json_rpc/index.js]: Looks great, thanks!

File: blockchain/block.js (hunk 7 above)
Comment (liayoo) [reply blockchain/block.js]: Done!

File: db/functions.js (hunk 8 above)
Comment (liayoo) [reply db/functions.js]: Done!

File: blockchain/block.js (hunk 7 above)
Comment (liayoo) [reply blockchain/block.js]: Actually, let me update this to the owner's address as discussed offline.

Comment (liayoo) [reply blockchain/block.js]: Done!

File: common/constants.js [hunk 9]
```diff
@@ -55,6 +54,7 @@ const CHAINS_H2N_DIR_NAME = 'h2n'; // Note: Block hash to block number
 const HASH_DELIMITER = '#';
 const TX_NONCE_ERROR_CODE = 900;
 const TX_TIMESTAMP_ERROR_CODE = 901;
+const AI_NETWORK = 'AI_Network';
```
Comment (platfowner) [inline common/constants.js:57]: We don't need this anymore?

Comment (liayoo) [reply common/constants.js]: Done!

**Review by platfowner — APPROVED** [review]
LGTM

Oh, I was thinking that this PR was already approved. Sorry for this late approve.
Nice job done. Thanks!

**Comment by liayoo** [comment]
No worries! I was actually caught up in something else and also thought this was already merged 😄  Thanks for the review! Now merging..

**Pull request merged**